crates/cow-amm/src/maintainers.rs

```diff
@@ -0,0 +1,87 @@
+use {
+    crate::cache::Storage,
+    contracts::ERC20,
+    ethcontract::{errors::MethodError, futures::future::join_all, Address},
+    ethrpc::Web3,
+    shared::maintenance::Maintaining,
+    std::{
+        collections::{HashMap, HashSet},
+        sync::Arc,
+    },
+    tokio::sync::RwLock,
+};
+
+pub struct EmptyPoolRemoval {
+    storage: Arc<RwLock<Vec<Storage>>>,
+    web3: Web3,
+}
+
+impl EmptyPoolRemoval {
+    pub fn new(storage: Arc<RwLock<Vec<Storage>>>, web3: Web3) -> Self {
+        Self { storage, web3 }
+    }
+
+    /// Checks if the given AMM has a zero balance of the specified token.
+    async fn check_single(
```

Comment:
ok, well like this, this method doesn't really save any reader's complexity. I meant more of a `check_single(&self, amm: AMM)` which joins all the checks for the individual tokens (this avoids one level of nesting in the main method).

Reply:
Got it. https://github.com/cowprotocol/services/pull/2849/commits/3401b9575b57eb6681ec4109a6ff4c7086497561